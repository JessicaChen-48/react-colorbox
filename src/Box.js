import "./Box.css";

function Box({ width, height, backgroundColor, removeBox}) {
  let boxDetails = { width, height, backgroundColor };
  return (
    <div>
      <div className="box" style={boxDetails}>
      </div>
      <button onClick={removeBox}>REMOVE THE BOX</button>
    </div>
  );
}

export default Box;
