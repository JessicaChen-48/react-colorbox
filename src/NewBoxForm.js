import {useState} from "react";

function NewBoxForm({handleForm}) {

  let initialData = {
    width:"",
    height:"",
    backgroundColor: ""
  };

  const [formData, setFormData] = useState(initialData);

  function handleChange(e) {
    e.preventDefault();
    const { name, value } = e.target;
    setFormData(formData => ({
      ...formData,
      [name]: value,
    }));
  }

  function handleSubmit(e) {
    e.preventDefault();
    let {height, width} = formData;

    let intHeight = parseInt(height);
    let intWidth = parseInt(width);

    if (intHeight && intWidth) {
      handleForm(formData);
      setFormData(initialData);
    } else {
      if(!intHeight) {
        setFormData(formData => ({...formData, height:"Invalid Height!"}));
      }
      if(!intWidth) {
        setFormData(formData => ({...formData, width:"Invalid Width!"}));
      }
    }
  }

  return (
  <form onSubmit={handleSubmit}>
    <label htmlFor="width">Width: </label>
    <input
      id="width"
      name="width"
      value={formData.width}
      onChange={handleChange}/>

    <label htmlFor="height">Height: </label>
      <input
        id="height"
        name="height"
        value={formData.height}
        onChange={handleChange}/>

    <label htmlFor="backgroundColor">Background Color: </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        value={formData.backgroundColor}
        onChange={handleChange} />
    <button>SUBMIT</button>
  </form>
  );
}

export default NewBoxForm;