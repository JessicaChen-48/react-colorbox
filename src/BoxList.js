import {useState} from "react"
import {v4 as uuid} from "uuid"
import Box from "./Box.js"
import NewBoxForm from "./NewBoxForm"

function BoxList() {

  const [boxes, setBoxes] = useState([{width:"100px", height:"100px", backgroundColor:"red"}]);

  function removeBox(idx) {
    let copy = boxes.map(box => ({...box}))
    setBoxes(copy.filter((c, i) => i !== idx))
  }

  function handleForm({width, height, backgroundColor}) {
    width = `${width}px`
    height = `${height}px`
    let copy = boxes.map(box => ({...box}))
    setBoxes([...copy, {width, height, backgroundColor}])
  }

  return (
    <>
    {boxes.map((box, i) =>
      <Box 
        width={box.width}
        height={box.height}
        backgroundColor={box.backgroundColor}
        removeBox={() => removeBox(i)}
        key = {uuid()}
        />)}
    <NewBoxForm handleForm={handleForm}/>
    </>
  )
}


export default BoxList;