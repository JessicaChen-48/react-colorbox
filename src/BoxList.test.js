import React from "react";
import { render, fireEvent, debug } from "@testing-library/react";
import BoxList from "./BoxList";
import Box from "./Box.js";
import NewBoxForm from "./NewBoxForm.js"



it("Box smoke test", function () {
  render(<Box />);
});

it("Box snapshot test", function() {
  let box = { width: "100px",
    height: "100px",
    backgroundColor: "black"
  };

  const {container} = render(<Box {...box}/>);
  expect(container).toMatchSnapshot();

});

it("NewBoxForm smoke test", function () {
  render(<NewBoxForm />)
});

it("NewBoxForm snapshot test", function() {
  const {container} = render(<NewBoxForm />);
  expect(container).toMatchSnapshot();
});

it("BoxList smoke test", function () {
  render(<BoxList />)
});

it("BoxList snapshot test", function() {
  const {container} = render(<BoxList />);
  expect(container).toMatchSnapshot();
});

it("NewBoxForm correctly changes and submits form", function() {
  const {container, getByLabelText, queryByText, debug} = render(<BoxList />);

  expect(container.querySelector(".box")).not.toBeInTheDocument();

  const widthInput = getByLabelText('Width:');
  const heightInput = getByLabelText('Height:');
  const colorInput = getByLabelText('Background Color:');
  const submitButton = queryByText('SUBMIT');

  fireEvent.change(widthInput, {target: {value: "100"}});
  fireEvent.change(heightInput, {target: {value: "200"}});
  fireEvent.change(colorInput, {target: {value: "green"}});

  expect(widthInput.value).toEqual("100");
  expect(heightInput.value).toEqual("200");
  expect(colorInput.value).toEqual("green");

  fireEvent.click(submitButton);
  expect(container.querySelector(".box")).toBeInTheDocument();


  expect(container.querySelector(".box").getAttribute('style'))
    .toEqual("width: 100px; height: 200px; background-color: green;");
});

it("NewBoxForm validates height and width on submit", function() {
  const {container, getByLabelText, queryByText} = render(<BoxList />);

  const widthInput = getByLabelText('Width:');
  const heightInput = getByLabelText('Height:');
  const colorInput = getByLabelText('Background Color:');
  const submitButton = queryByText('SUBMIT');

  fireEvent.change(widthInput, {target: {value: "hello"}});
  fireEvent.change(heightInput, {target: {value: "bye"}});
  fireEvent.change(colorInput, {target: {value: "green"}});

  fireEvent.click(submitButton);

  expect(widthInput.value).toEqual("Invalid Width!");
  expect(heightInput.value).toEqual("Invalid Height!");
  expect(colorInput.value).toEqual("green");
});


it("BoxList automatically makes steelblue color box if no color submitted", function() {
  const {container, getByLabelText, queryByText} = render(<BoxList />);

  const widthInput = getByLabelText('Width:');
  const heightInput = getByLabelText('Height:');
  const colorInput = getByLabelText('Background Color:');

  const submitButton = queryByText('SUBMIT');

  fireEvent.change(widthInput, {target: {value: "100"}});
  fireEvent.change(heightInput, {target: {value: "100"}});
  fireEvent.change(colorInput, {target:{value:"fdalskjfakl;ds"}});

  fireEvent.click(submitButton);

  expect(container.querySelector(".box").getAttribute('style'))
  .toEqual("width: 100px; height: 100px;");

})

it("BoxList properly removes box if button clickd", function() {
  const {container, getByLabelText, queryByText} = render(<BoxList />);

  const widthInput = getByLabelText('Width:');
  const heightInput = getByLabelText('Height:');
  const colorInput = getByLabelText('Background Color:');

  const submitButton = queryByText('SUBMIT');

  fireEvent.change(widthInput, {target: {value: "100"}});
  fireEvent.change(heightInput, {target: {value: "200"}});
  fireEvent.change(colorInput, {target: {value: "green"}});

  fireEvent.click(submitButton);

  const removeButton = queryByText('REMOVE THE BOX');

  fireEvent.click(removeButton);

  expect(container.querySelector(".box")).not.toBeInTheDocument();

})



