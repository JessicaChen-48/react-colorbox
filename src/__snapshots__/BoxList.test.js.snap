// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box snapshot test 1`] = `
<div>
  <div>
    <div
      class="box"
      style="width: 100px; height: 100px; background-color: black;"
    />
    <button>
      REMOVE THE BOX
    </button>
  </div>
</div>
`;

exports[`BoxList snapshot test 1`] = `
<div>
  <div>
    <form>
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        value=""
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        value=""
      />
      <label
        for="backgroundColor"
      >
        Background Color: 
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        value=""
      />
      <button>
        SUBMIT
      </button>
    </form>
  </div>
  <br />
</div>
`;

exports[`NewBoxForm snapshot test 1`] = `
<div>
  <form>
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      name="width"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      name="height"
      value=""
    />
    <label
      for="backgroundColor"
    >
      Background Color: 
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      value=""
    />
    <button>
      SUBMIT
    </button>
  </form>
</div>
`;
